File: PA117.java
//Leetcode : 42. Trapping Rain Water
//Time Complexity: O(n)
//Space Complexity:O(1)
class Solution {
    public int trap(int[] height) {
        int lmax=0;
        int rmax=0;
        int water=0;
        int i=0;
        int j=height.length-1;
        while(i<j){
            if(height[i]<height[j]){
                 if(lmax<height[i]){
                    lmax= height[i];
                }
                else{
                    water= water+(lmax-height[i]);
                }
                i++;
            }
            else{
                 if(rmax<height[j]){
                    rmax= height[j];
                }
                else{
                    water= water+(rmax-height[j]);
                }
                j--;
            }
           
        }
        
      return water;  
    }
}

File: PA117.py
#Time Complexity - O(n)
#Space Complexity - O(1)
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        
        hm = 0
        lm = 0
        l = 0
        r = len(height) - 1
        ans = 0
        while l < r:
            if height[l] < height[r]:
                if height[l] > lm:
                    lm = height[l]
                else:
                    ans = ans + (lm - height[l])
                l = l + 1  
            else:
                if height[r] > hm:
                    hm = height[r]
                else:
                    ans = ans + (hm - height[r])
                r = r - 1
        return ans